src/client.rs

```diff
@@ -165,6 +165,26 @@ impl Client {
             trace!(?response, "Response received");
             match response {
                 Ok(resp) => {
+                    // Check HTTP status code first before parsing body
+                    if let Err(err) = resp.error_for_status_ref() {
+                        if err.is_status() {
+                            let status_code = err.status().unwrap();
+                            return Err(match status_code {
+                                StatusCode::UNAUTHORIZED => ClientError::Status(
+                                    401,
+                                    "Unauthorized - check RPC username/password".to_string(),
+                                ),
+                                StatusCode::FORBIDDEN => {
+                                    ClientError::Status(403, "Forbidden".to_string())
+                                }
+                                StatusCode::NOT_FOUND => {
+                                    ClientError::Status(404, "Not Found".to_string())
+                                }
+                                _ => ClientError::Status(status_code.as_u16(), err.to_string()),
+                            });
+                        }
+                    }
```

Comment:
How about we get the canonical reason for status code so that we don't have to write "Forbidden", "Not found" etc ourselves. That will also handle cases for other status codes automatically. Something like this:

```rust
let resp = match resp.error_for_status() {
    Err(e) if e.is_status() => {
        let status = e.status().unwrap();
        let reason =
            status.canonical_reason().unwrap_or(e.to_string());
        return Err(ClientError::Status(status.into(), reason.to_string()));
    }
    Err(e) => {
        return Err(ClientError::Other(e.to_string()));
    }
    Ok(resp) => resp,
};
```

I hope `status.into()` works.

Reply:
Doing this should also completely address #23